internal/service/deleter.go

```diff
@@ -0,0 +1,109 @@
+package service
+
+import (
+	"log"
+	"sync"
+	"time"
+
+	"github.com/mkukarin01/snort/internal/storage"
+)
+
+// deleteJob - структурка тасков на удаления
+type deleteJob struct {
+	userID   string
+	shortIDs []string
+}
+
+// URLDeleter - структурка для фанин, по ней мы пачкой "удалим" записи
+type URLDeleter struct {
+	store   storage.Storager
+	inChan  chan deleteJob
+	stopCh  chan struct{}
+	wg      sync.WaitGroup
+	bufSize int
+}
+
+// NewURLDeleter - создаём новый агрегатор
+func NewURLDeleter(store storage.Storager) *URLDeleter {
+	return &URLDeleter{
+		store:   store,
+		inChan:  make(chan deleteJob, 100), // буфер
+		stopCh:  make(chan struct{}),
+		bufSize: 100, // сброс/флеш
+	}
+}
+
+// Run запускатор, цикл на чтение и агрегация тасков по фанин
+func (d *URLDeleter) Run() {
+	// какой-то странный костыль, когда запускаешь горутину, если она ничего не сделала, а ты выполнил Stop
+	// произойдет `panic: sync: negative WaitGroup counter` с ненулевым wg этой проблемы нет
+	d.wg.Add(1)
+
+	flushInterval := time.Second
+
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	// локальная мапа для агрегации
+	userBatch := make(map[string][]string)
+
+	flushFunc := func() {
+		if len(userBatch) == 0 {
+			return
+		}
+		d.flush(userBatch)
+		// чистим
+		userBatch = make(map[string][]string)
+	}
+
+	for {
+		select {
+		case <-d.stopCh:
+			// финалочка
+			flushFunc()
+			d.wg.Done()
+			return
+		case job := <-d.inChan:
+			// копим
+			userBatch[job.userID] = append(userBatch[job.userID], job.shortIDs...)
+
+			// при пороге - надо провести сброс
+			var count int
+			for _, ids := range userBatch {
+				count += len(ids)
+			}
```

Comment:
```suggestion
			count += len(job.shortIDs)
```
в текущем исполнении получается почти O(n2). Можно вынести определение счетчика на один уровень с userBatch и обнулять их одновременно в flushFunc. Не блокер.